ReplicatedMap fetchEntryViews API [HZ-3932]

srknzl opened the pull request:
Mono PR: https://github.com/hazelcast/hazelcast-mono/pull/188


This PR extends #489

ihsandemir commented on copy_java_files.sh:
```diff
@@ -0,0 +1,11 @@
+# Use this to copy generated java files to hazelcast repo all at once. 
```
generator.py already has an option where you can specify destination, why do you need this?

srknzl replied:
yeah apparently it has, I didn't notice. I'll delete the script

ihsandemir commented on protocol-definitions/ReplicatedMap.yaml:
```diff
@@ -820,3 +820,113 @@ methods:
+    since: 2.7
+    doc: |
+      Copies all of the mappings from the specified entry list to this map, including metadata.
+      This method uses ReplicatedRecordStore.putRecords method with version=current version + 1.
```
What do you mean by `current version`? How does it work when it is concurrent invocations?

srknzl replied:
putRecords is used to migrate all records from one member to another member, every store has some current version that is increased for every modification (I believe). That's why we need the version information here

srknzl replied:
Btw the description is wrong here, we should be just putting records with the provided version. Let me fix here

srknzl replied:
Probably left stale because I based on another protocol PR in this PR

ihsandemir commented on protocol-definitions/ReplicatedMap.yaml:
```diff
@@ -820,3 +820,113 @@ methods:
+    since: 2.7
+    doc: |
+      Allows iteration over ReplicatedMapEntryViewHolder objects. The changes happened during the iteration may not be included in the iterated EntryViews. This method
+      will throw an exception if there is no replicated record store with the given replicated map name and partition id. This method will consume some memory in the 
```
how much is this `some memory`? If I have many concurrent requests how is memory usage growing?

srknzl replied:
Depends on the entry sizes and fetchSize parameter.

ihsandemir commented on protocol-definitions/ReplicatedMap.yaml:
```diff
@@ -852,3 +852,75 @@ methods:
+      will throw an exception if there is no replicated record store with the given replicated map name and partition id. This method will consume some memory in the 
+      member with the default timeout of 300 seconds that is reset after each fetch. Sending endEntryViewIteration or timing out will release the resource.
+    request:
+      retryable: true
```
You say that this API causes some memory allocation and keeps it at server memory. What happens to that memory on retry since you made it retryable?

frant-hartm replied:
The memory per client iteration is constant.
If a request is retried with same cursor id it doesn't allocate new memory.

ihsandemir commented on protocol-definitions/ReplicatedMap.yaml:
```diff
@@ -852,3 +852,75 @@ methods:
+          nullable: true
+          since: 2.7
+          doc: |
+            The identifier of the last fetched page. Initially null and then returned by this API. 
```
How do you guarantee the uniqueness of the UUID? Let's assume I am doing many `fetchEntryViews ` calls(may be million).

frant-hartm replied:
It's a UUID, it should be a non-issue.

ihsandemir commented on protocol-definitions/ReplicatedMap.yaml:
```diff
@@ -852,3 +852,75 @@ methods:
+    response:
+      params:
+        - name: cursorId
+          type: UUID
```
is this different from request parameter `cursorId`? Why?

frant-hartm replied:
This is the cursor id for the *next* batch.

srknzl replied:
You keep feeding the response cursor id to request cursor id to fetch the new pages. It's a simple id of the current page.

ihsandemir commented on protocol-definitions/ReplicatedMap.yaml:
```diff
@@ -852,3 +852,75 @@ methods:
+          nullable: false
+          since: 2.7
+          doc: |
+            A unique string which is used to fetch new pages.
```
it is not `String`!!! UUID

srknzl replied:
a UUID is unique so I don't need to say unique UUID. So a UUID is basically can be represented by a string, that's what I meant there. anyway

ihsandemir commented on protocol-definitions/ReplicatedMap.yaml:
```diff
@@ -852,3 +852,75 @@ methods:
+          nullable: true
+          since: 2.7
+          doc: |
+            A list of EntryViews. Null means end of iteration. 
```
Does this mean that the client always needs to make the last fetch call which has to return null? Can we avoid this extra call? E.g. fi returned size is less than the requested batch size, it means, the entries are finished.

srknzl replied:
no you can't do that if the page size divides the total entry count evenly.

ihsandemir commented on protocol-definitions/ReplicatedMap.yaml:
```diff
@@ -852,3 +852,75 @@ methods:
+    name: endEntryViewIteration
+    since: 2.7
+    doc: |
+      This method is used to release the resource generated by fetchEntryViews. Send this request after you retrieved the last page and no
```
why do we need the client to make this request? Server knows it finished and it should release the resources. What is the rationale to make server cleanup depend on a client request?

frant-hartm replied:
We need this extra request to tell the server that the client has received the last page of data.

If we don't do this the following could happen:
- client sends a request - it is the last page
- server clears the cursor
- response with the last page gets lost
- client retries, but server doesn't have the data anymore.

However, I don't think we need a new request type for this, we can send another request when client receives batch < max batch size (or empty), the client knows the response to this will be empty, so it doesn't even need to wait for that 2nd response.

srknzl replied:
If page size divides entry count evenly, that logic won't work. I can add another parameter "endIteration:true" to the original method but I prefer a new method because it resembles the SQL close statement.

srknzl replied:
I think by empty you mean null page. it can work but I don't see a problem in a new method

srknzl commented:
@ahmetmircik @frant-hartm Apologies but I had to stale your reviews and change the design a bit after dealing with test failures. I've added the new "newIteration" request parameter that should be set to true if requesting creation of a new iteration. It can be safely retried if set to true, if there is a iteration with the cursor id, it won't create a new iteration. 

This mainly fixes the problem if a client sends a null cursor id (previous design) and does not receive the cursor id somehow. Then, it'll retry the request and create another completely different iterator. That's why the initial uuid is created on the client side now. 

I've also noticed considerable amount of bugs and flaky tests and fixed them all in the main pr.

srknzl commented:
Another thing I've ended up doing was one of your suggestions in the past. The response page is not nullable anymore and if received items in the page is less than the batchSize, the iteration ends. Note that client is responsible for clearing resources even if the iteration ends. A member only cleans it via the cleanup future

ihsandemir reviewed (approved):
Approved based on other approvals.
